gcp/gce_instance.go

```diff
@@ -0,0 +1,85 @@
+package gcp
+
+import (
+	"errors"
+	"strings"
+	"time"
+)
+
+func zoneFromUrl(url string) (string, error) {
+	split := strings.Split(url, "/")
+	if len(split) == 0 {
+		return "", errors.New("got invalid zone url: " + url)
+	}
+	return split[len(split)-1], nil
+}
+
+func regionFromZone(ctx *GcpContext, zone string) (string, error) {
+	for _, region := range ctx.Regions {
+		for _, regionZoneUrl := range region.Zones {
+			regionZone, err := zoneFromUrl(regionZoneUrl)
+			if err != nil {
+				return "", nil
+			}
+			if zone == regionZone {
+				return region.Name, nil
+			}
+		}
+	}
+
+	return "", errors.New("could not get region for zone: " + zone)
+}
+
+func GetAllGceInstances(ctx *GcpContext, excludedRegions []string, excludeAfter time.Time) ([]GcpResource, error) {
+	instances := []GcpResource{}
+
+	apiInstances, err := ctx.Service.Instances.AggregatedList(ctx.Project).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, item := range apiInstances.Items {
+		for _, apiInstance := range item.Instances {
+			// skip if deletion protection is turned on
+			if apiInstance.DeletionProtection {
+				continue
+			}
+
+			zone, err := zoneFromUrl(apiInstance.Zone)
+			if err != nil {
+				return nil, err
+			}
+
+			region, err := regionFromZone(ctx, zone)
+			if err != nil {
+				return nil, err
```

Comment:
Consider logging errors and collecting them in a list, but not returning them until the very end. That way, a single failed instance delete early on doesn't prevent all others from being deleted later.

Reply:
It sounds like this comment was meant for the nuke code. I updated the nuke code to log errors and keep going.